File: rust/examples/ws_book_and_trades.rs
mod common;

use common::{market_and_orderbook, other, rest_client, ExampleResult};
use futures_util::StreamExt;
use lightcone::prelude::*;
use tokio::time::{timeout_at, Duration, Instant};

#[tokio::main]
async fn main() -> ExampleResult {
    let client = rest_client()?;
    let (_, orderbook) = market_and_orderbook(&client).await?;
    let mut ws = client.ws_native();
    let orderbook_id = orderbook.orderbook_id.clone();

    // One connection can hold multiple aggregation views of the same book.
    // Key the local state by the frame's aggregation: full precision for
    // pricing, a grouped view (5 sig figs, mantissa 2) for display.
    let grouped_aggregation =
        BookAggregation::validate(Some(5), Some(2)).map_err(|message| other(message))?;
    let mut full_book = OrderbookState::new(orderbook_id.clone());
    let mut grouped_book =
        OrderbookState::with_aggregation(orderbook_id.clone(), grouped_aggregation);
    let mut trades = TradeHistory::new(orderbook_id.clone(), 20);

    ws.connect().await?;
    ws.send(MessageOut::subscribe_books(
        vec![orderbook_id.clone()],
        BookAggregation::FULL,
    ))?;
    ws.send(MessageOut::subscribe_books(
        vec![orderbook_id.clone()],
        grouped_aggregation,
    ))?;
    ws.subscribe(SubscribeParams::Trades {
        orderbook_ids: vec![orderbook_id.clone()],
    })?;

    let mut hits = 0;
    {
        let events = ws.events();
        tokio::pin!(events);

        let deadline = Instant::now() + Duration::from_secs(30);
        while hits < 4 {
            let Ok(Some(event)) = timeout_at(deadline, events.next()).await else {
                println!("no more websocket data (timeout or stream ended)");
                break;
            };

            match event {
                WsEvent::Message(Kind::BookUpdate(update)) => {
                    // Untagged frames are the full-precision view; frames from
                    // the grouped subscription carry n_sig_figs/mantissa.
                    let aggregation = update.aggregation();
                    let book = if aggregation == grouped_aggregation {
                        &mut grouped_book
                    } else {
                        &mut full_book
                    };
                    if update.resync {
                        // Refresh exactly the affected view: re-subscribe with
                        // the SAME aggregation and reset its revision gate.
                        ws.send(MessageOut::unsubscribe_books(
                            vec![update.id.clone()],
                            aggregation,
                        ))?;
                        book.begin_generation();
                        ws.send(MessageOut::subscribe_books(
                            vec![update.id.clone()],
                            aggregation,
                        ))?;
                        continue;
                    }
                    book.apply(&update);
                    println!(
                        "book[{}]: seq={} bid={:?} ask={:?}",
                        aggregation.key_suffix(),
                        book.seq,
                        book.best_bid(),
                        book.best_ask()
                    );
                    hits += 1;
                }
                WsEvent::Message(Kind::Trade(trade)) => {
                    println!(
                        "trade: {} {} @ {} seq={}",
                        trade.size, trade.side, trade.price, trade.sequence
                    );
                    trades.push(trade.into());
                    hits += 1;
                }
                WsEvent::Connected => {
                    full_book.begin_generation();
                    grouped_book.begin_generation();
                }
                WsEvent::Error(err) => eprintln!("ws error: {err}"),
                _ => {}
            }
        }
    }

    ws.disconnect().await?;
    if hits == 0 {
        return Err(other("received no websocket events — connection may be broken").into());
    }
    println!("buffered trades: {}", trades.len());
    Ok(())
}

File: typescript/examples/ws_book_and_trades.ts
import {
  aggregationFromFrame,
  aggregationKeySuffix,
  aggregationsEqual,
  OrderbookState,
  subscribeBooks,
  TradeHistory,
  unsubscribeBooks,
  validateAggregation,
  type BookAggregation,
  type Trade,
  type WsEvent,
} from "../src";
import { marketAndOrderbook, restClient, runExample, withTimeout } from "./common";

async function main() {
  const client = restClient();
  const [, orderbook] = await marketAndOrderbook(client);
  const orderbookId = orderbook.orderbookId;

  // One connection can hold multiple aggregation views of the same book.
  // Key the local state by the frame's aggregation: full precision for
  // pricing, a grouped view (5 sig figs, mantissa 2) for display.
  const groupedAggregation: BookAggregation = validateAggregation({ nSigFigs: 5, mantissa: 2 });
  const fullBook = new OrderbookState(orderbookId);
  const groupedBook = new OrderbookState(orderbookId);
  const trades = new TradeHistory(orderbookId, 20);
  const ws = client.ws();
  let hits = 0;

  let resolveDone!: () => void;
  const done = new Promise<void>((resolve) => {
    resolveDone = resolve;
  });

  const unsubscribe = ws.on((event: WsEvent) => {
    if (event.type === "Message" && event.message.type === "book_update") {
      const update = event.message.data;
      // Untagged frames are the full-precision view; frames from the grouped
      // subscription carry n_sig_figs/mantissa.
      const aggregation = aggregationFromFrame(update.n_sig_figs, update.mantissa);
      if (update.resync) {
        // Refresh exactly the affected view: re-subscribe with the SAME
        // aggregation. The fresh snapshot arrives with seq 0 and replaces
        // the book (last-write-wins).
        ws.send(unsubscribeBooks([update.orderbook_id], aggregation));
        ws.send(subscribeBooks([update.orderbook_id], aggregation));
        return;
      }
      const book = aggregationsEqual(aggregation, groupedAggregation) ? groupedBook : fullBook;
      book.apply(update);
      console.log(
        `book[${aggregationKeySuffix(aggregation)}]: seq=${book.seq} bid=${book.bestBid()} ask=${book.bestAsk()}`
      );
      hits += 1;
    } else if (event.type === "Message" && event.message.type === "trades") {
      const trade: Trade = {
        orderbookId: event.message.data.orderbook_id,
        tradeId: event.message.data.trade_id,
        timestamp: new Date(event.message.data.timestamp),
        price: event.message.data.price,
        size: event.message.data.size,
        side: event.message.data.side,
        sequence: event.message.data.sequence,
      };
      console.log(`trade: ${trade.size} ${trade.side} @ ${trade.price} seq=${trade.sequence}`);
      trades.push(trade);
      hits += 1;
    } else if (event.type === "Error") {
      console.error("ws error:", event.error);
    }

    if (hits >= 4) {
      resolveDone();
    }
  });

  try {
    await ws.connect();
    ws.subscribe({ type: "book_update", orderbook_ids: [orderbookId] });
    ws.subscribe({
      type: "book_update",
      orderbook_ids: [orderbookId],
      nSigFigs: groupedAggregation.nSigFigs,
      mantissa: groupedAggregation.mantissa,
    });
    ws.subscribe({ type: "trades", orderbook_ids: [orderbookId] });
    await withTimeout(done, 30_000, "timed out waiting for websocket data");
  } catch {
    console.log("no more websocket data (timeout or stream ended)");
  } finally {
    unsubscribe();
    await ws.disconnect();
  }

  if (hits === 0) {
    throw new Error("received no websocket events — connection may be broken");
  }
  console.log(`buffered trades: ${trades.len()}`);
}

void runExample(main);
